Switch to using NUnit Test Framework

jthelin opened the pull request:
Switch over to using the `NUnit` test framework (http://nunit.net/) for Project Orleans.

More explanation of the reasons behind this switch are posted in #574

If you have any general comments or suggestions about changing test framework, then please reply to that thread.

Please use this PR thread for any specific feedback about the code-level changes themselves.

jthelin commented:
Squashed working branch and rebased against GH master branch @ 5b914ccfeaca07424c543845a24f076dff16b5a1

jthelin commented:
No feedback received on this change proposal, so are we ok to merge this to master?

gabikliot commented:
I personally haven't had a chance yet to look deeply and try it. I think we want to wait with merging untill 1.0.9 is released.

sergeybykov commented:
Yep, that's why it is targeted to 1.0.10.

jthelin commented:
I mistakenly though all those code signing jobs we did last week _were_ 1.0.9 so we were done, but apparently not with more this morning ;)

sergeybykov commented:
Last week for just the VSIX. :-)

veikkoeeva commented:
Is it OK if I just ask just here if it still possible to run the SQL tests on Windows as-is? E.g. the script is found, loaded to the databaset etc. automatically?

gabikliot commented:
@jthelin, I just talked to @jdom (one of our big internal customers) and he raised a valid concern regarding moving to NUnit. If our existing customers (internal and external) are using MSTest and for some reason cannot move to NUnit (either already have too large test code base or using some MSTest feature which does not exist in NUnit), this will present an issue.
On the one hand they can in theory still test their grains even with in memory silos in MSTest, while we do our own testing in NUnit. But on the there hand, it means they won't be able to use our TesrtSiloHOst as is. Even if they could, the whole procedure of testing will be deferent- different deployment attributes, different places to deploy and copy configs, etc...
So if we don't provide a direct support for MSTest, it will make their life harder to use MSTest.

I am not saying this concern should necessarily block our progress with this PR, but we should definitely give it a thought.

jthelin commented:
By design, OrleansTestingHost has NO dependencies on any particular testing framework.
It does not include any assembly references to either NUnit or MsTest in its package / csproj.
It is just a POCO class as far as .NET is concerned.

Other projects should still be able to use OrleansTestingHost with whichever test framework they want / are current using. 

I don't think there will be any changes required to the current Functional and 343 test suites either.

The only point of contact with a particular test framework It should be the the project containing the test cases themselves, which obviously will need to use the appropriate test annotation attributes for the framework they use.

NUnit itself installs cleanly as a NuGet package, so there should be zero additional install footprint on any build / test lab servers either.

Hope this addresses your concerns, or happy to skype chat if you want to.

sergeybykov commented:
@jthelin We can still run SQL tests against the DB file, can't we?

gabikliot commented:
Let me explain better what I meant. Sure, OrleansTestingHost has no dependencies on any particular testing framework, that is good. What I meant is that our usage of it has a dependency on a concrete test framework. We have dependency in the order of test, init, cleanup, ... The complicated dependency in MSTest  was the primary reason why you suggested NUnit in the first place. So if we now pick to escape this complication and use NUNit, of course other would be able to write their own tests in MSTTests, but they will be left on their own dealing with MSTest intricate complexities of init order. And that may set them to failure. Specifically, they will not be able to take our tests as is, and run thee as is, or use them as an example in MSTest. So although we wont prevent them from using MSTest with OrleansTestingHost, we also won't help them in any way. On the other hand, if we stay with MSTest, we "eat our own dog food". Again, not saying we should not, just need to be aware of the consequences.

jthelin commented:
I understand the words you are saying @gabikliot but i don't know why or what they mean :(

For example [just to name two MS projects are random], how is this any different than which test framework CoreClr Project chooses to use (it is not MsTest, btw) or which one WCF uses (it is not MsTest either) 

To me which test framework other teams choose to use, and how they choose to run those test suites seems completely orthogonal to which one __we__ use internally on Orleans GH project.
We don't claim to support other project's test suites / labs, do we? 
So why is other people's test suites suddenly our problem?

This PR does not change __any__ Orleans runtime code. 

So the Orleans core runtime code that other teams are already using today in there test suites is completely unchanged by this change set. 

The only thing that is changing here is **our** tests, which are primarily for our use, no?

I'm sorry, i just don't understand how it is even possible to address the hypothetical concerns you seem to have.

P.S. I am ranked #8 in my building, #461 in Redmond and #2036 in Microsoft overall on my MSElite internal dogfood programs participation leaderboard, so i already know exactly what "eat our own dogfood" is all about! I already do my part ;)

jthelin commented:
Rebased against master HEAD @ 1de0fe2cd8dac79e6025a173adb07fb2860e5fc4

jthelin commented:
Rebased against master HEAD @ 04749c4c7e6cdb70ff8d315308a353aa0680ba5b

veikkoeeva commented:
@jthelin It appears that it is not possible to run SQL Server or Azure Table Storage tests automatically anymore. Do you plan to add these still? I would be in favour of doing this with an added twist that we check the platfom and run them only on Windows.

With regard to relational storage, my intention was to clean up a bit the initialization logic and arrange it so that it checks default ports of SQL Server, MySQL etc., compares them to scripts there are in Orleans and runs the tests automatically. Currently just only for SQL Server as that's the only relational backend having SQL scripts. The goal of this is to make adding and checking relational backends as easy as possible.

gabikliot commented:
We run automatic tests with azure table every night in our internal CI framework. It works.

veikkoeeva commented:
@gabikliot Hey, thanks for chiming in! I was just investigating a bit if this is due to me having Azure SDK 2.7 installed now. And indeed, I see the scripts being used in the code. But I wonder why the tests get stuck on my computer.

@jthelin This was a false alarm on my part. Sorry for that. I'll investigate later on the week further.

veikkoeeva commented:
Taking a very quick look, getting stuck at ``silo.Start();`` in ``StartOrleansSilo``, say, when running ``Liveness_Sql_1``. Might be something from my part, I merely cloned the upstream branch. If this pollutes the thread, let's remove this (but now I need to get sleep). VS 2015 RTM, Windows 8.1, Azure SDK 2.7 (shouldn't matter on SQL tests).

```
te.processhost.managed.exe Error: 0 : [2015-07-21 19:50:23.227 GMT    81	ERROR  	100071	Dispatcher	127.0.0.1:22224]	!!!!!!!!!! SelectTarget failed with Unexpected: Cannot find an implementation class for grain interface 0	
Exc level 0: Orleans.Runtime.OrleansException: Unexpected: Cannot find an implementation class for grain interface 0
   kohteessa Orleans.Runtime.GrainTypeManager.GetTypeInfo(Int32 typeCode, String& grainClass, PlacementStrategy& placement, String genericArguments) tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\GrainTypeManager\GrainTypeManager.cs:rivillä 146
   kohteessa Orleans.Runtime.Catalog.GetGrainTypeInfo(Int32 typeCode, String& grainClass, PlacementStrategy& placement, String genericArguments) tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Catalog\Catalog.cs:rivillä 317
   kohteessa Orleans.Runtime.Placement.PlacementContextExtensions.GetGrainTypeName(IPlacementContext this, Int32 typeCode, String genericArguments) tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Placement\IPlacementContext.cs:rivillä 85
   kohteessa Orleans.Runtime.Placement.PlacementContextExtensions.GetGrainTypeName(IPlacementContext this, GrainId grainId, String genericArguments) tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Placement\IPlacementContext.cs:rivillä 91
   kohteessa Orleans.Runtime.Placement.RandomPlacementDirector.OnAddActivation(PlacementStrategy strategy, GrainId grain, IPlacementContext context) tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Placement\RandomPlacementDirector.cs:rivillä 62
   kohteessa Orleans.Runtime.Placement.PlacementDirectorsManager.AddActivation(GrainId grain, IPlacementContext context, PlacementStrategy strategy) tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Placement\PlacementDirectorsManager.cs:rivillä 111
   kohteessa Orleans.Runtime.Placement.PlacementDirectorsManager.<SelectOrAddActivation>d__11.MoveNext() tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Placement\PlacementDirectorsManager.cs:rivillä 90
--- Pinon jäljityksen loppu aiemmasta sijainnista, jossa palautettiin poikkeus ---
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter`1.GetResult()
   kohteessa Orleans.Runtime.Dispatcher.<AddressMessage>d__31.MoveNext() tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Core\Dispatcher.cs:rivillä 555
--- Pinon jäljityksen loppu aiemmasta sijainnista, jossa palautettiin poikkeus ---
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   kohteessa System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   kohteessa Orleans.Runtime.Dispatcher.<AsyncSendMessage>d__29.MoveNext() tiedostossa C:\projektit\jthelin\orleans\src\OrleansRuntime\Core\Dispatcher.cs:rivillä 514
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
Exception thrown: 'System.AggregateException' in mscorlib.dll
te.processhost.managed.exe Information: 0 : [2015-07-21 19:50:23.315 GMT   106	INFO   	100632	MembershipFactory	127.0.0.1:22224]	-Waiting for membership table provider to initialize. Going to sleep for 00:05:00 and re-try to reconnect.	
te.processhost.managed.exe Information: 0 : [2015-07-21 19:50:48.892 GMT    43	INFO   	100507	Catalog	127.0.0.1:22223]	Before collection#3: memory=24MB, #activations=0, collector=<#Activations=0, #Buckets=0, buckets=[]>.	
te.processhost.managed.exe Information: 0 : [2015-07-21 19:50:48.893 GMT    23	INFO   	100507	Catalog	127.0.0.1:22222]	Before collection#3: memory=24MB, #activations=1, collector=<#Activations=0, #Buckets=0, buckets=[]>.	
te.processhost.managed.exe Information: 0 : [2015-07-21 19:50:48.898 GMT    43	INFO   	100508	Catalog	127.0.0.1:22223]	After collection#3: memory=24MB, #activations=0, collected 0 activations, collector=<#Activations=0, #Buckets=0, buckets=[]>, collection time=00:00:00.0063327.	
te.processhost.managed.exe Information: 0 : [2015-07-21 19:50:48.901 GMT    23	INFO   	100508	Catalog	127.0.0.1:22222]	After collection#3: memory=24MB, #activations=1, collected 0 activations, collector=<#Activations=0, #Buckets=0, buckets=[]>, collection time=00:00:00.0083598.	

```

gabikliot commented:
>. If this pollutes the thread, let's remove this

I agree. This is unrelated to @jthelin NUnit PR. Lets handle it separately.

jthelin commented:
Rebased against master HEAD @ c5d698435cb5a114a53352909c4525b1ce0acec0

jthelin commented:
The SQL tests run OK for me if you set "shadow copy" option in your NUnit test runner [eg ReSharper]

There are some remaining problems with the more complex Liveness tests, but those repro with MembershipGrain too, so I think it is a small bug in the base class. In this change set I did not change the Liveness test cases, so they are still commented out like they are in master.

jthelin commented:
Rebased against master HEAD @ d01468db339a48a0a63356b1e5cd7e4089f78946

jthelin commented:
Rebased against master HEAD @ 9abdf192751b2685a5bf2e01942a567792f2f08f

veikkoeeva commented:
@jthelin Out of curiosity, is this slated for the next release?

jthelin commented:
Yes, it is still on the cards. 

End-of-Internship time is super busy at Microsoft Research :(

I just need to steal some CPU cycles to update the change set and sync the timing of the switch-over with @sergeybykov who is currently enjoying a well deserved vacation in Hawaii :sunglasses:

veikkoeeva commented:
@jthelin I'll add here one view that CoreCLR is centering around [xUnit.NET](http://blog.stephencleary.com/2015/03/continuous-integration-code-coverage-open-source-net-coreclr-projects.html). Maybe not that important otherwise than that on Gitter there's been views xUnit.NET would be better. There's always differing opinions/desires. :), This is well argued, though, so just recording here it's been seen.

jthelin commented:
Updated and rebased against GH master branch @ 9c57a23

jthelin commented:
@veikkoeeva - Thanks for the link to Stephen Cleary's blog post. Very informative. 

While I am aware of various opinions and reasonings that xUnit is "better" then the other available test frameworks, unfortunately I don't have spare cycles to convert things over to anything other than NUnit right now.

No pushback from me if someone else wants to take on the job of converting to xUnit though ;)

jthelin commented:
@sergeybykov - I believe this is ready to merge, if you would like to take a final look.

sergeybykov commented on src/Tester/MembershipTests/LivenessTests.cs:
```diff
@@ -216,77 +244,54 @@ public class LivenessTests_MembershipGrain : LivenessTestsBase
-            DoClassInitialize();
-        }
-
-        //[TestMethod, TestCategory("BVT"), TestCategory("Functional"), TestCategory("Membership"), TestCategory("Gabi")]
```
These tests were commented out. Presumably because they were broken. Do they work now?

sergeybykov commented:
Wouldn't this change kill CI? I see 'Build finished. No test results found.' in the CI build results and Console Output seems to have a bunch of exceptions despite reported success.

Locally, I see a strange behavior. Test.cms ran fine with 95 tests succeeded, but when I try to run the tests with ReSharper in VS 2013, they are failing. Am I missing something?

jthelin commented:
@sergeybykov
> when I try to run the tests with ReSharper in VS 2013, they are failing. 

You need to ensure that the "Shadow copy" option is turned off in ReSharper
https://www.jetbrains.com/resharper/help/Reference__Options__Tools__Unit_Testing.html

The `Test.cmd` script uses the `/noshadow` option to do that automatically.

By default, ReSharper shadow copies .dll binaries, but not config files.

veikkoeeva commented:
@jthelin Heh, that is true. A good position to take that of yours. Whilst searching for opinions on [a paper](https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-yuan.pdf) I linked to Gitter, I came across [http://ayende.com/blog/171809/repeatable-random-tests](
Repeatable random tests) by *Oren Eini*. Maybe I should do some research on what exactly separates NUnit from Xunit.NET. Perhaps also, what's the real deal with IntelliTest, Code Contracts, FsCheck etc. It's a bit difficult that some of these go deep into maths I can't handle and I don't have the time to learn. :)

sergeybykov commented:
@jthelin Yep, disabling of Shadow copy got the tests running. Thanks. What about CI?

jthelin commented:
For CI builds, in "theory" it should be just a case of tweaking some build params / steps to use NUnit `.xml` format test result files rather VsTest `.trx` format test result files. 

I took a look at that yesterday, and of course things are a little more complex in practice! 

There is almost a complete lack of any good docs for the build process / steps, plus Visual Studio complicates the situation right now by doing some auto-publish magic through the .testsettings file at the moment.

Now that PR #743 got rid of the `TestInput` directory [which was almost exclusively what the Local.testsettings file was used for previously, modulo the hidden auto-publish side-effect] 
it should be easier to tease apart the test execution and test results collection steps in the build 
and plug in NUnit results collection instead.

Further complication is we do builts with at least three different systems currently: 
Jenkins-CI [external / GitHub] and VSO [internal] and Appveyer [personal]. 
Arrrgh!

Continuing to explore solution in my 20% time ........... [aka nights and weekends!]

veikkoeeva commented:
@jthelin Do allocate family time too (other than this great open source family)! :)

Interestingly, I just read from the MS blogs [Supplying Run Time Parameters to Tests](http://blogs.msdn.com/b/visualstudioalm/archive/2015/09/04/supplying-run-time-parameters-to-tests.aspx) and first comment, I feel, is spot on. Other way is transforming the ``.config`` files with XDT.

sergeybykov commented:
Should we close this one for now?

jthelin commented:
Yes, i'll close this for now. It looks like will need to clone the current Jenkins build job config and play around with settings. There is a post build option to publish NUnit test results files, but need to prototype where they are located on the build servers before this change could be turned on.

sergeybykov commented:
@veikkoeeva on Gitter pointed to the following.
>@sergeybykov, @galvesribeiro To get NUnit or xUnit.NET to work on VSO (or what's it's called nowdays), one needs to do some additional copying: http://stackoverflow.com/questions/24888661/unable-to-get-visual-studio-online-to-execute-my-nunit-tests.
Maybe stating the obvious, but I want to state anyway. :smile:
http://listofsoftwareontfshostedbuildserver.azurewebsites.net/
